lib/Dialect/XTenNN/IR/XTenNNOps.cpp

```diff
@@ -593,3 +597,39 @@ bool TopK::isCompatibleReturnTypes(mlir::TypeRange l, mlir::TypeRange r) {
       getElementTypeOrSelf(l[1]) == getElementTypeOrSelf(r[1]);
   return sameElementType && succeeded(verifyCompatibleShapes(l, r));
 }
+
+LogicalResult ReduceMeanOp::inferReturnTypeComponents(
+    MLIRContext * /*context*/, std::optional<Location> location,
+    ReduceMeanOp::Adaptor adaptor,
+    SmallVectorImpl<ShapedTypeComponents> &inferredReturnShapes) {
+
+  auto inTy = cast<RankedTensorType>(adaptor.getInput().getType());
+  auto inDims = inTy.getShape();
+  auto keepDims = adaptor.getKeepdims();
+
+  auto axes = adaptor.getAxes();
+  llvm::SmallVector<int64_t> newAxes(inDims);
+  for (auto axis : axes) {
+    // onnx spec: axis: [-r, r-1]
+    if (axis < -inTy.getRank() || axis >= inTy.getRank()) {
+      return emitOptionalError(location,
+                               "expected axis to be within [-rank,rank) (where "
+                               "rank is the rank of the input)");
+    }
+    // normalize axis: [0, r)
+    if (axis < 0) {
+      axis += inTy.getRank();
+    }
+    assert((axis >= 0 && axis < inTy.getRank()) && "axis has invalid value");
+
+    if (keepDims) {
+      newAxes[axis] = 1;
+    } else {
+      newAxes.erase(newAxes.begin() + axis);
```

Comment:
This is going to go out of bound in the next iteration.

Reply:
I think it'd be better to go over the dims of input and if it's among the axis, push back to `inDims` only if keepdims through.

Reply:
Fixed it.